### PR TITLE
support get_worker() and worker_client() in async tasks

### PR DESCRIPTION
**Pull request by graingert**
Closes https://github.com/dask/distributed/issues/5513

- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by graingert** [comment]
this will also need a test for `get_worker()` in Actors

**Comment by graingert** [comment]
this is like https://github.com/dask/distributed/pull/5517 but only sets the contextvar when running user tasks

File: distributed/client.py [hunk 1]
```diff
@@ -2579,10 +2579,11 @@ def scatter(
...
             )
 
-        if hasattr(thread_state, "execution_state"):  # within worker task
-            local_worker = thread_state.execution_state["worker"]
-        else:
+        try:
+            local_worker = get_worker()
+        except ValueError:
             local_worker = None
+
```
Comment (graingert) [inline distributed/client.py:2586]:
```suggestion
```

File: distributed/worker_client.py [hunk 2]
```diff
@@ -53,22 +53,26 @@ def worker_client(timeout=None, separate_thread=True):
...
-                key=thread_state.key,
-                compute_duration=duration,
-                stimulus_id=f"worker-client-secede-{time()}",
-            ),
-        )
-
-    yield client
-
-    if separate_thread:
-        rejoin()
```
Comment (graingert) [inline distributed/worker_client.py:71]: previously this didn't secede if there was an exception in `with worker_client()`

**Review by hendrikmakait — APPROVED** [review]
Thanks, @graingert!

**Pull request merged**